src/distributions/float.rs

```diff
@@ -64,6 +91,66 @@ macro_rules! float_impls {
                 fraction.into_float_with_exponent(0) - (1.0 - EPSILON / 2.0)
             }
         }
+
+        impl Distribution<$ty> for HighPrecision01 {
+            /// Generate a floating point number in the half-open interval
+            /// `[0, 1)` with a uniform distribution. See [`HighPrecision01`].
+            ///
+            /// # Algorithm
+            /// (Note: this description used values that apply to `f32` to
+            /// illustrate the algorithm).
+            ///
+            /// The trick to generate a uniform distribution over [0,1) is to
+            /// set the exponent to the -log2 of the remaining random bits. A
+            /// simpler alternative to -log2 is to count the number of trailing
+            /// zeros in the random bits. In the case where all bits are zero,
+            /// we simply generate a new random number and add the number of
+            /// trailing zeros to the previous count (up to maximum exponent).
+            ///
+            /// Each exponent is responsible for a piece of the distribution
+            /// between [0,1). We take the above exponent, add 1 and negate;
+            /// thus with probability 1/2 we have exponent -1 which fills the
+            /// range [0.5,1); with probability 1/4 we have exponent -2 which
+            /// fills the range [0.25,0.5), etc. If the exponent reaches the
+            /// minimum allowed, the floating-point format drops the implied
+            /// fraction bit, thus allowing numbers down to 0 to be sampled.
+            ///
+            /// [`HighPrecision01`]: struct.HighPrecision01.html
+            #[inline]
+            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
+                // Unusual case. Separate function to allow inlining of rest.
+                #[inline(never)]
+                fn fallback<R: Rng + ?Sized>(mut exp: i32, fraction: $uty, rng: &mut R) -> $ty {
+                    // Performance impact of code here is negligible.
+                    let bits = rng.gen::<$uty>();
+                    exp += bits.trailing_zeros() as i32;
+                    // If RNG were guaranteed unbiased we could skip the
+                    // check against exp; unfortunately it may be.
+                    // Worst case ("zeros" RNG) has recursion depth 16.
```

Comment:
The recursion can be removed easily. The contents of `fallback` can look like the code below, which I also find a little easier to understand.

```rust
// Worst case ("zeros" RNG) will have 16 iterations (f64 exp bias / 64).
loop {
    let bits = rng.gen::<$uty>();
    exp += bits.trailing_zeros() as i32;
    
    // if exp reaches bias, return a subnormal that can be zero
    if exp >= $exponent_bias {
        return fraction.into_float_with_exponent(-$exponent_bias);
    }
    if bits != 0 {
        return fraction.into_float_with_exponent(-exp);
    }
}
```

Reply:
Actually I just thought the recursion was neater since @pitdicker recommended this bit be in a separate function anyway (not inlined so that the common code is small). Since it will almost never recurse I don't see a problem with that strategy. But using a loop also works (was used in an older commit).